### PR TITLE
Implement relationship discovery for OGR layers, and show relationships in browser

### PR DESCRIPTION
**Pull request by nyalldawson**
(requires GDAL 3.6 or later!)

This PR implements two user-facing changes:

- it allows the Project Properties - Relations - Discover Relations feature to work for OGR layers which support relationships. Specifically, this includes GeoPackage (via the related tables extension) and ESRI FileGeodatabases
- it shows detailed relationship information for these data sources in the browser panel (see screenshot below)
 
image

I've included a bunch of related cleanups to the relationship classes in QGIS. There's two main parts to this:

1. A general cleanup including moving common enums to qgis.h and promoting to enum classes
2. An expansion of the QgsWeakRelation class, so that it is able to losslessly represent the full details of relationships available from the GDAL API (which was designed to be a "superset" of relationship properties from common formats)

The docs give a good breakdown on why this was added to QgsWeakRelation:

```
 * In constrast to QgsRelation, QgsWeakRelation can be used to encapsulate
 * information about a relationship which does not currently exist in a QGIS project.
 * E.g. it can be used to represent a relationship which exists in a database
 * backend (but not within a QGIS project). Accordingly, some properties
 * available in QgsWeakRelation are included for informational purposes only,
 * and cannot be translated to QgsRelations or respected in QGIS relationships.
```

So the primary distinction here is that a QgsWeakRelation represents a "theoretical" relationship from a datasource, whereas QgsRelation is the (lossy) functional version of that relationship in a QGIS project.

### CODE REVIEW

**Comment by nyalldawson** [comment]
Note that a follow up PR will implement new logic which auto-adds related tables when a parent table is added from an OGR source (and automatically creates the corresponding relationship)

File: src/core/providers/ogr/qgsogrproviderconnection.cpp [hunk 1]
```diff
@@ -858,4 +863,177 @@ QgsAbstractDatabaseProviderConnection::SqlVectorLayerOptions QgsOgrProviderConne
...
+    }
+    return output;
+  }
+  else
+  {
+    throw QgsProviderConnectionException( QObject::tr( "There was an error opening the dataset %1!" ).arg( uri() ) );
+  }
+#else
+  Q_UNUSED( tableName )
+  throw QgsProviderConnectionException( QObject::tr( "Retrieving relationships for datasets requires GDAL 3.5 or later" ) );
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrproviderconnection.cpp:1035]:
```suggestion
  throw QgsProviderConnectionException( QObject::tr( "Retrieving relationships for datasets requires GDAL 3.6 or later" ) );
```

File: src/app/qgsrelationadddlg.cpp [hunk 2]
```diff
@@ -141,17 +139,17 @@ void QgsRelationAddDlg::updateFieldsMappingHeaders()
   mFieldsMappingTable->setVerticalHeaderLabels( verticalHeaderLabels );
 }
 
-QString QgsRelationAddDlg::referencingLayerId()
+QString QgsRelationAddDlg::referencingLayerId() const
```
Comment (nirvn) [inline src/app/qgsrelationadddlg.cpp:142]: This makes me think we should harmonize our use of Dlg vs Dialog class name in QGIS 4.0.

In the meantime, might be worth killing the _only_ use of dlg in src/app, that is QgsRelationAddDlg :) doesn't have to happen in this PR.

File: src/core/browser/qgsrelationshipsitem.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        catch ( QgsProviderConnectionException &ex )
+        {
+          relationError = ex.what();
+        }
+
+        for ( const QgsWeakRelation &relation : std::as_const( relations ) )
+        {
+          try
+          {
+            QgsRelationshipItem *relationshipItem { new QgsRelationshipItem( this, relation ) };
```
Comment (nirvn) [inline src/core/browser/qgsrelationshipsitem.cpp:65]:
```suggestion
            QgsRelationshipItem *relationshipItem = new QgsRelationshipItem( this, relation );
```

File: src/core/providers/ogr/qgsgeopackageproviderconnection.cpp [hunk 4]
```diff
@@ -336,6 +336,10 @@ void QgsGeoPackageProviderConnection::setDefaultCapabilities()
   mCapabilities |= Capability::ListFieldDomains;
 #endif
 
+#if GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(3,6,0)
```
Comment (nirvn) [inline src/core/providers/ogr/qgsgeopackageproviderconnection.cpp:339]: This might well kick me into building gdal locally instead of relying on the distro package :)

File: src/core/qgis.h [hunk 5]
```diff
@@ -2072,6 +2072,48 @@ class CORE_EXPORT Qgis
...
 
+    /**
+     * Relationship types.
+     *
+     * \note Prior to QGIS 3.28 this was available as QgsRelation::RelationType.
+     *
+     * \since QGIS 3.28
+     */
+    enum class RelationshipType SIP_MONKEYPATCH_SCOPEENUM_UNNEST( QgsRelation, RelationType ) : int
+      {
```
Comment (nirvn) [inline src/core/qgis.h:2083]: Weird that the auto-formatter didn't catch this one.

```suggestion
{
```

File: src/core/qgis.h [hunk 6]
```diff
@@ -2072,6 +2072,48 @@ class CORE_EXPORT Qgis
...
+
+    /**
+     * Relationship strength.
+     *
+     * \note Prior to QGIS 3.28 this was available as QgsRelation::RelationStrength.
+     *
+     * \since QGIS 3.28
+     */
+    enum class RelationshipStrength SIP_MONKEYPATCH_SCOPEENUM_UNNEST( QgsRelation, RelationStrength ) : int
+      {
```
Comment (nirvn) [inline src/core/qgis.h:2097]:
```suggestion
{
```

File: src/core/qgsogrutils.cpp [hunk 7]
```diff
@@ -998,6 +998,8 @@ QgsFields QgsOgrUtils::stringToFields( const QString &string, QTextCodec *encodi
 QStringList QgsOgrUtils::cStringListToQStringList( char **stringList )
 {
   QStringList strings;
+  if ( !stringList )
```
Comment (nirvn) [inline src/core/qgsogrutils.cpp:1001]: What about moving that _above_ the QStringList strings to save a few CPU cycles? :)

File: src/core/qgspolymorphicrelation_p.h [hunk 8]
```diff
@@ -30,11 +30,12 @@
 // version without notice, or even be removed.
 //
 
-#include "qgspolymorphicrelation.h"
-#include "qgsrelation.h"
-
 #include <QSharedData>
 #include <QPointer>
+#include "qgis.h"
```
Comment (nirvn) [inline src/core/qgspolymorphicrelation_p.h:35]: Can we move QGIS headers to site above the Qt ones?

File: src/core/qgsweakrelation.cpp [hunk 9]
```diff
@@ -17,64 +17,196 @@
...
+    case Qgis::RelationshipCardinality::OneToMany:
+    case Qgis::RelationshipCardinality::ManyToOne:
+    {
+
+      QgsRelation relation;
+      relation.setId( mRelationId );
+      relation.setName( mRelationName );
+      relation.setStrength( mStrength );
+      QgsVectorLayerRef referencedLayerRef { mReferencedLayer };
+      QgsMapLayer *referencedLayer { referencedLayerRef.resolveWeakly( project, matchType ) };
```
Comment (nirvn) [inline src/core/qgsweakrelation.cpp:52]: There's a bunch of these which if it was up to me I'd re-write as QgsMapLayer *l = ...; instead. Feel free to ignore.

File: src/core/qgsweakrelation.cpp [hunk 10]
```diff
@@ -17,64 +17,196 @@
...
 {
-  return mFieldPairs;
+  return mReferencedLayer.provider;
+}
+
+QString QgsWeakRelation::referencedLayerName() const
+{
+  if ( QgsProviderMetadata *metadata = QgsProviderRegistry::instance()->providerMetadata( mReferencedLayer.provider ) )
+  {
+    return metadata->decodeUri( mReferencedLayer.source ).value( QStringLiteral( "layerName" ) ).toString();
```
Comment (nirvn) [inline src/core/qgsweakrelation.cpp:173]: I love how we just keep using this decodeUri more, and more, and more :)

File: tests/src/core/testqgsweakrelation.cpp [hunk 11]
```diff
@@ -66,72 +67,172 @@ void TestQgsWeakRelation::cleanup()
...
                            QStringLiteral( "memory" ),
                            QStringLiteral( "referencedLayerId" ),
                            QStringLiteral( "referencedLayerName" ),
                            QStringLiteral( "Polygon?crs=epsg:4326&field=pk:int&field=province:int&field=municipality:string" ),
-                           QStringLiteral( "memory" ),
-                           fieldPairs
+                           QStringLiteral( "memory" )
                          );
-  QVERIFY( ! weakRel.resolvedRelation( QgsProject::instance(), QgsVectorLayerRef::MatchType::Name ).isValid() );
+  weakRel.setReferencingLayerFields( {"fk_province" } );
```
Comment (nirvn) [inline tests/src/core/testqgsweakrelation.cpp:82]:
```suggestion
  weakRel.setReferencingLayerFields( { "fk_province" } );
```

File: tests/src/core/testqgsweakrelation.cpp [hunk 12]
```diff
@@ -66,72 +67,172 @@ void TestQgsWeakRelation::cleanup()
...
+  weakRel.setReferencedLayerFields( {"pk" } );
```
Comment (nirvn) [inline tests/src/core/testqgsweakrelation.cpp:83]: Auto-formatter should pick those up if possible.

```suggestion
  weakRel.setReferencedLayerFields( { "pk" } );
```

File: tests/src/core/testqgsweakrelation.cpp [hunk 13]
```diff
@@ -66,72 +67,172 @@ void TestQgsWeakRelation::cleanup()
...
+                           QStringLiteral( "referencingLayerId" ),
+                           QStringLiteral( "referencingLayerName" ),
+                           QStringLiteral( "Point?crs=epsg:4326&field=pk:int&field=fk_province:int&field=fk_municipality:int" ),
+                           QStringLiteral( "memory" ),
+                           QStringLiteral( "referencedLayerId" ),
+                           QStringLiteral( "referencedLayerName" ),
+                           QStringLiteral( "Polygon?crs=epsg:4326&field=pk:int&field=province:int&field=municipality:string" ),
+                           QStringLiteral( "memory" )
+                         );
+  weakRel.setReferencingLayerFields( {"fk_province" } );
```
Comment (nirvn) [inline tests/src/core/testqgsweakrelation.cpp:224]:
```suggestion
  weakRel.setReferencingLayerFields( { "fk_province" } );
```

File: tests/src/core/testqgsweakrelation.cpp [hunk 14]
```diff
@@ -66,72 +67,172 @@ void TestQgsWeakRelation::cleanup()
...
+  weakRel.setReferencedLayerFields( {"pk" } );
```
Comment (nirvn) [inline tests/src/core/testqgsweakrelation.cpp:225]:
```suggestion
  weakRel.setReferencedLayerFields( { "pk" } );
```

**Review by nirvn — APPROVED** [review]
Lovely stuff.

File: src/app/qgsrelationadddlg.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/app/qgsrelationadddlg.cpp:142]: Ok, rename to QgsCreateRelationDialog

File: src/core/qgis.h (hunk 5 above)
Comment (nyalldawson) [reply src/core/qgis.h:2083]: Ah, it always does that when the monkey patch macro is in use. Not much we can do there...

**Comment by nyalldawson** [comment]
Thanks @nirvn, all addressed!

**Comment by phidrho** [comment]
> Note that a follow up PR will implement new logic which auto-adds related tables when a parent table is added from an OGR source (and automatically creates the corresponding relationship)

Hi @nyalldawson,
I just updated GDAL to 3.6.0, and noticed automatic creation of relationship and adding child tables. I have noticed some things related to this but I'm not sure if they are bugs or is this behavior by design:

mtm example

1) I created an example with many-to-many relationship and it confused me why it doesn't work other way around - when adding child features why doesn't it automatically add parent features and sets relationship in project configuration (in case of many-to-many relationship).

2) Other thing I noticed that I get `WARNING` errors `Missing Relation in configuration` for mapping table when identifying features - either on child or parent feature.
warning message on every identify event

3) In `Browser` there is a ***empty*** field listed on mapping table with `Type: geometry` that actually doesn't exist in GPKG
empty layer

 I can open bug report with data and demo project if needed, but I wasn't sure if these are bugs but are by design (expected).

**Comment by nyalldawson** [comment]
@phidrho looks like your gpkg relationship doesn't follow the gpkg related tables specifications correctly. Can you share so that i can investigate? How was the gpkg relationship created?

**Comment by phidrho** [comment]
> @phidrho looks like your gpkg relationship doesn't follow the gpkg related tables specifications correctly. Can you share so that i can investigate? How was the gpkg relationship created?

@nyalldawson


here is the raw gpkg - created with QGIS - [raw_data.zip](https://github.com/qgis/QGIS/files/10010386/raw_data.zip),
and here are queries for adding relations:

# GPKG many-to-many relations test

## 1. I created some demo layers via QGIS GUI and stored them in GPKG "raw_data.gpkg"

- 2 related layers (both of type **features**)
  - my_points
  - my_polygons

related polygons and points

## 2. Manually added information about relations via SQL

- I followed [OGC GeoPackage Related Tables Extension](https://docs.ogc.org/is/18-000/18-000.html)

### 2.1 Added relation schema by defined in extension

#### 2.1.1 Defining extension

```sql
INSERT INTO gpkg_extensions (table_name, column_name, extension_name, definition, scope) VALUES
('relmap_points_polys', NULL, 'related_tables', 'TBD', 'read-write');
```

#### 2.1.2 Creating and populating relations table

```sql
CREATE TABLE 'gpkgext_relations' (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    base_table_name TEXT NOT NULL,
    base_primary_column TEXT NOT NULL DEFAULT 'id',
    related_table_name TEXT NOT NULL,
    related_primary_column TEXT NOT NULL DEFAULT 'id',
    relation_name TEXT NOT NULL,
    mapping_table_name TEXT NOT NULL UNIQUE
);
```

#### 2.1.2 Creating and populating relations data

I followed many-to-many example from [Annex B](https://docs.ogc.org/is/18-000/18-000.html#_example_informative):

```sql
CREATE TABLE relmap_points_polys (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    base_id INTEGER CONSTRAINT [fk_relmap_points_polys-my_polygons]
        REFERENCES my_polygons (fid)
        ON DELETE CASCADE
        ON UPDATE CASCADE
        DEFERRABLE INITIALLY DEFERRED,
    related_id INTEGER CONSTRAINT [fk_relmap_points_polys-my_points]
        REFERENCES my_points (fid)
        ON DELETE CASCADE
        ON UPDATE CASCADE
        DEFERRABLE INITIALLY DEFERRED,
    created_on INTEGER DEFAULT (strftime('%s'))
);


CREATE INDEX [idx_relmap_points_polys-base_id] ON relmap_points_polys (
    base_id
);


CREATE INDEX [idx_relmap_points_polys-related_id] ON relmap_points_polys (
    related_id
);


INSERT INTO gpkgext_relations (base_table_name, base_primary_column, related_table_name, related_primary_column, relation_name, mapping_table_name) VALUES
('my_polygons', 'fid', 'my_points', 'fid', 'features', 'relmap_points_polys');
```

Schema visualisation - as described in gpkgext_relations - same as in Annex B:

schema

### 2.2 Adding GPKG metadata about relation table

```sql
INSERT INTO
    gpkg_contents (
        table_name,
        data_type,
        identifier,
        description,
        last_change,
        srs_id
    )
VALUES
    (
        'relmap_points_polys',
        'attributes',
        'relmap_points_polys',
        'Relation table between points and polygons',
        strftime('%Y-%m-%dT%H:%M:%fZ', 'now'),
        '0'
    );


INSERT INTO
    gpkg_ogr_contents (table_name, feature_count)
VALUES
    ('relmap_points_polys', '0');


CREATE TRIGGER trigger_delete_feature_count_relmap_points_polys
    AFTER
        DELETE ON relmap_points_polys BEGIN
    UPDATE
        gpkg_ogr_contents
    SET
        feature_count = feature_count - 1
    WHERE
        lower(table_name) = lower('relmap_points_polys');
END;


CREATE TRIGGER trigger_insert_feature_count_relmap_points_polys
    AFTER
        INSERT ON relmap_points_polys BEGIN
    UPDATE
        gpkg_ogr_contents
    SET
        feature_count = feature_count + 1
    WHERE
        lower(table_name) = lower('relmap_points_polys');
END;
```

## 3. Adding some relational data

```sql
INSERT INTO relmap_points_polys (base_id,related_id) VALUES
(1,1),
(2,1),
(1,2),
(2,2),
(1,3),
(1,4),
(2,5),
(2,6);
```

**Comment by nyalldawson** [comment]
@phidrho 

I'm not sure what I'm doing different, but the db created using your process works fine for me!

I've attached it here -- can you spot the difference?


[created.zip](https://github.com/qgis/QGIS/files/10099484/created.zip)

**Comment by nyalldawson** [comment]
@phidrho 
> In Browser there is a empty field listed on mapping table with Type: geometry that actually doesn't exist in GPKG

This is fixed by https://github.com/qgis/QGIS/pull/51028 (it's harmless though -- it just shows at the moment for all non-spatial tables in gpkg)

**Comment by phidrho** [comment]
> @phidrho
> 
> I'm not sure what I'm doing different, but the db created using your process works fine for me!
> 
> I've attached it here -- can you spot the difference?
> 
> [created.zip](https://github.com/qgis/QGIS/files/10099484/created.zip)

Hi @nyalldawson,

first of all, thank you for checking this.

Everything works for me also, it's nothing wrong with implementation in code but in logic. I think it should be upgraded to recognize relationship in both ways on `many-to-many` relationships.

There are two cases:
1) when I add new point I need to reference it to specific polygon/s
  - notes:
      - or none, if polygon/s currently doesn't exist - this means no data entry to `mapping table` in the moment of creating a point, but should be possible afterwards
      - this is OK because `points` and `polygons` tables don't reference mapping table by themself, there is only FK from `mapping table` to `points` and `polygons` table

2) when I add new polygon, I need to reference points that belong to it
  - same notes from above apply here

Currently software recognizes only one of these cases, depends how you define relation.

I'm demonstrating this by adding another relation to GPKG (this is obviously wrong approach, because we have two mapping tables that we need to keep in sync by ourselves, but just for demo it's OK), with additional mapping table:
[raw_data_both_direction.zip](https://github.com/qgis/QGIS/files/10101806/raw_data_both_direction.zip)
Now it works as expected, you can add `points` or `polygons` layer and relation is recognized.

So, long story short, when `many-to-many` relation is detected, it should automatically add both way relations to QGIS project, and when either of these two layers are added, other layer and mapping table should be also added to project.
It's in the middle of a feature request and bug report, certain logic is missing but nothing is wrong with it, that's why I posted it here in first place.

I can understand if currently there's no funding for it, we can always add this other direction manually in the meantime, but then we should open new feature request proposal for this scenario.

And we also still have these *WARNING messages*, at least on my side, not sure if they come from GDAL or QGIS - also reported here on PostGIS data - #48776

**Comment by nyalldawson** [comment]
@phidrho 

> So, long story short, when many-to-many relation is detected, it should automatically add both way relations to QGIS project,

This should already be happening -- eg when I open my version of the raw_data my_polygons table, I get the following relationships auto created in the QGIS project:

image

Is this what you see?

I should point out that there ARE some rough edges with many-to-many relationships in QGIS, which were out of scope for this work. The goal here was simply to get the relationships setup automatically and unlock the existing QGIS relationship support (which is fantastic for 1 to many, slightly fragile for many-to-many!)

**Comment by phidrho** [comment]
@nyalldawson 
> Is this what you see?

Yes it's same on my side, but only if I load `my_polygons` from GPKG, then both `mapping table` and `my_points` layers are also loaded and relations are set up as expected.

But if I load only `my_points` layer to empty project, none of these two relationships are defined, nor `my_polygons` and `mapping table` layers gets loaded automatically. It's a many-to-many relationship (which is correctly recognized in Browser panel) so I expected it to also work in this case.

With this example from my previous comment: [raw_data_both_direction.zip](https://github.com/qgis/QGIS/files/10101806/raw_data_both_direction.zip), it works either if you try to load only `my_points` or `my_polygons` to empty project - everything gets set up as expected (relations and other layers).
